### PR TITLE
Ordering satisfy consider ordering equivalence of different lengths

### PR DESCRIPTION
**Pull request by berkaysynnada**
# Which issue does this PR close?

Closes #6280.

# Rationale for this change

`ordering_satisfy_concrete()` function first checks whether the required PhysicalSortExpr's length is greater than provided PhysicalSortExpr's length. If so, it directly returns false. However, there may be cases such as:

Provided: (a)
Required: (a,b) ,but a and b are in an equivalent class.

In that case, an unnecessary sort is added. We can remove it by changing the implementation of `ordering_satisfy_concrete()`.

# What changes are included in this PR?

Required PhysicalSortExpr's are normalized, and simplified to have unique elements. Provided PhysicalSortExpr's are also normalized before the comparison of expressions. The length check is moved to the final, after the normalization and compression steps. 

# Are these changes tested?

Yes, both in utils.rs and window.slt.

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/physical-expr/src/equivalence.rs [hunk 1]
```diff
@@ -153,7 +153,7 @@ impl<T: Eq + Hash + Clone> EquivalentClass<T> {
     }
 
     pub fn insert(&mut self, col: T) -> bool {
-        self.others.insert(col)
+        self.head != col && self.others.insert(col)
     }
```
Comment (ozankabak) [inline datafusion/physical-expr/src/equivalence.rs:157]: I ran into this silent bug when investigating another issue (which has its own comment below). Please check if my change is correct.

Comment (ozankabak) [inline datafusion/physical-expr/src/equivalence.rs:157]: I ran into this silent bug when investigating another issue (which has its own comment below). Please check if my change is correct.

File: datafusion/physical-expr/src/equivalence.rs [hunk 2]
```diff
@@ -168,7 +168,7 @@ impl<T: Eq + Hash + Clone> EquivalentClass<T> {
                 false
             }
         } else {
-            true
+            removed
```
Comment (ozankabak) [inline datafusion/physical-expr/src/equivalence.rs:171]: Same as above, please verify.

**Review by ozankabak — COMMENTED** [review]
I was stepping over the code in debug mode to understand how it works, and I may have come across there may be a silent bug in the way we handle ordering-equivalences. To understand the issue, first note that `normalize_sort_expr_with_ordering_equivalence_properties` and `normalize_sort_requirement_with_ordering_equivalence_properties` share the same implementation (the former uses the latter). However, is this really what we should be doing?

The possible issue arises when the normalized expression is *not equal* to the original. In that case, we simply use the sort options for the normalized expression and ignore the sort options of the original. This may be OK for the former case (i.e. for sort expressions), but is it OK with sort requirements too? Options of the given requirement is independent of what we have in our equivalence classes, and we need to respect it regardless.

Please think about this with @mustafasrepo and we can discuss altogether in our next meeting. If this is indeed a bug and you check in a fix, please include a unit test to exercise this as well.

**Comment by mustafasrepo** [comment]
Merged upstream

**Pull request closed without merging**